src/lib.rs

```diff
@@ -76,28 +86,372 @@ impl ULinkZenoh {
+    async fn send_response(
+        &self,
+        zenoh_key: &str,
+        payload: UPayload,
+        attributes: UAttributes,
+    ) -> Result<(), UStatus> {
+        // Get the data from UPayload
+        let Some(Data::Value(buf)) = payload.data else {
+            // TODO: Assume we only have Value here, no reference for shared memory
+            return Err(UStatus::fail_with_code(
+                UCode::InvalidArgument,
+                "Invalid data",
+            ));
+        };
+
+        // Serialized UAttributes into protobuf
+        // TODO: Should we map priority into Zenoh priority?
+        let mut attr = vec![];
+        let Ok(()) = attributes.encode(&mut attr) else {
+            return Err(UStatus::fail_with_code(
+                UCode::InvalidArgument,
+                "Unable to encode UAttributes",
+            ));
+        };
+        // Get reqid
+        let reqid = ULinkZenoh::uuid_to_string(&attributes.reqid.ok_or(
+            UStatus::fail_with_code(UCode::InvalidArgument, "reqid doesn't exist"),
+        )?);
+
+        // Add attachment and payload
+        let mut attachment = AttachmentBuilder::new();
+        attachment.insert("uattributes", attr.as_slice());
+        // Send back query
+        let value = Value::new(buf.into()).encoding(Encoding::WithSuffix(
+            KnownEncoding::AppCustom,
+            payload.format.to_string().into(),
+        ));
+        let reply = Ok(Sample::new(
+            KeyExpr::new(zenoh_key.to_string()).map_err(|_| {
+                UStatus::fail_with_code(UCode::Internal, "Unable to create Zenoh key")
+            })?,
+            value,
+        ));
+        let query = self
+            .query_map
+            .lock()
+            .unwrap()
+            .get(&reqid)
+            .ok_or(UStatus::fail_with_code(
+                UCode::Internal,
+                "query doesn't exist",
+            ))?
+            .clone();
+
+        // Send data
+        // TODO: Unable to use unwrap in with_attachment (Attachment doesn't have Debug trait)
+        if query
+            .reply(reply)
+            .with_attachment(attachment.build())
+            .map_err(|_| UStatus::fail_with_code(UCode::Internal, "Unable to add attachment"))?
+            .res()
+            .await
+            .is_err()
+        {
+            return Err(UStatus::fail_with_code(
+                UCode::Internal,
+                "Unable to reply with Zenoh",
+            ));
+        }
+
+        Ok(())
+    }
 }
 
 #[async_trait]
 impl RpcClient for ULinkZenoh {
     async fn invoke_method(
+        &self,
         topic: UUri,
         payload: UPayload,
         attributes: UAttributes,
     ) -> RpcClientResult {
-        // Do the validation
+        // Validate UUri
+        // TODO: Whether we should check URI is resolved or not
+        //       https://github.com/eclipse-uprotocol/uprotocol-spec/issues/42#issuecomment-1882273981
         if UriValidator::validate(&topic).is_err() {
```

Comment:
I suggest to use the more idiomatic `.map_err(|e| ...)?;` form.

Reply:
Make sense. I used to think in C way to handle the error. :satisfied: